### PR TITLE
WIP: aM+bN pauli sums

### PR DESCRIPTION
**Pull request by stylewarning**
This adds optimizations for exp(i(aM+bN)).

### CODE REVIEW

File: src/compilers/linear-paulis.lisp [hunk 1]
```diff
@@ -206,6 +281,20 @@
           (unless (= 1 (crawl-parameter (pauli-term-prefactor term)))
             (give-up-compilation))))
       
+      ;; test if it's one of the easy cases first.
+      (alexandria:when-let ((type (pauli-instr-of-type-aM+bN instr)))
```
Comment (notmgsk) [inline src/compilers/linear-paulis.lisp]: `a:when-let`

**Comment by stylewarning** [comment]
CC @ecpeterson

File: src/compilers/linear-paulis.lisp [hunk 2]
```diff
@@ -163,6 +163,81 @@
...
+           (and (= 2 (length words))
+                (every #'PP? words)
+                ;; Note that these interns are not
+                ;; free-for-alls. They'll only be II, XX, YY, or ZZ.
+                (let ((type0 (intern (first words) ':keyword))
+                      (type1 (intern (second words) ':keyword)))
+                  (and (not (eq type0 type1))
+                       (not (eq ':II type0))
+                       (not (eq ':II type1))
+                       (list type0 type1))))))))
```
Comment (notmgsk) [inline src/compilers/linear-paulis.lisp]: oof

File: src/compilers/linear-paulis.lisp [hunk 3]
```diff
@@ -163,6 +163,81 @@
                        (application-arguments instr))
                 (inst "CNOT" () control-qubit subvote-literal)))))))))
 
+;;; Helper functions for special case Paulis.
+
+(defun inst-aXX+bZZ (a b p q)
+  "Inst a program for aXX + bZZ on qubits p and q. This is
```
Comment (notmgsk) [inline src/compilers/linear-paulis.lisp]: INST

File: src/compilers/linear-paulis.lisp [hunk 4]
```diff
@@ -206,6 +281,20 @@
           (unless (= 1 (crawl-parameter (pauli-term-prefactor term)))
             (give-up-compilation))))
       
+      ;; test if it's one of the easy cases first.
```
Comment (ecpeterson) [inline src/compilers/linear-paulis.lisp]: I was going to recommend writing this as a separate `define-compiler`, but now that I see the gross-ass guard above this I understand your choice.

File: src/compilers/linear-paulis.lisp [hunk 5]
```diff
@@ -163,6 +163,81 @@
...
+  (error "idk"))
+
+(defun inst-aM+bN (a b p q m n)
+  (check-type m (member :XX :YY :ZZ))
+  (check-type n (member :XX :YY :ZZ))
+  (assert (not (eq m n)))
+  (optima:match (list m n)
+    ((list ':XX ':YY)
+     (inst-aXX+bYY a p q p)
+     )
```
Comment (notmgsk) [inline src/compilers/linear-paulis.lisp]: this one's trying to run away

File: src/compilers/linear-paulis.lisp [hunk 6]
```diff
@@ -163,6 +163,81 @@
...
+    ((list ':XX ':ZZ)
+     (inst-aXX+bZZ a b p q))
+    ((list ':YY ':ZZ)
+     (error "not implemented"))
+    (otherwise
+     (inst-aM+bN b a q p n m))))
+
+(defun pauli-instr-of-type-aM+bN (instr)
```
Comment (ecpeterson) [inline src/compilers/linear-paulis.lisp]: I might call this `…aMM+bNN` to set expectations.

File: src/compilers/linear-paulis.lisp (hunk 4 above)
Comment (stylewarning) [reply src/compilers/linear-paulis.lisp]: I also didn't know how the compilers would interact, esp. if this one succeeds.

**Pull request closed without merging**